libs/libskynet/src/skylinkBitfield.ts

```diff
@@ -0,0 +1,166 @@
+// parseSkylinkBitfield parses a skylink bitfield and returns the corresponding
+// version, offset, and fetchSize.
+function parseSkylinkBitfield(skylink: Uint8Array): [bigint, bigint, bigint, string | null] {
+  // Validate the input.
+  if (skylink.length !== 34) {
+    return [0n, 0n, 0n, "provided skylink has incorrect length"];
+  }
+
+  // Extract the bitfield.
+  let bitfield = new DataView(skylink.buffer).getUint16(0, true);
+
+  // Extract the version.
+  const version = (bitfield & 3) + 1;
+  // Only versions 1 and 2 are recognized.
+  if (version !== 1 && version !== 2) {
+    return [0n, 0n, 0n, "provided skylink has unrecognized version"];
+  }
+
+  // If the skylink is set to version 2, we only recognize the link if
+  // the rest of the bits in the bitfield are empty.
+  if (version === 2) {
+    if ((bitfield & 3) !== bitfield) {
+      return [0n, 0n, 0n, "provided skylink has unrecognized version"];
+    }
+    return [BigInt(version), 0n, 0n, null];
+  }
+
+  // Verify that the mode is valid, then fetch the mode.
+  bitfield = bitfield >> 2;
+  if ((bitfield & 255) === 255) {
+    return [0n, 0n, 0n, "provided skylink has an unrecognized version"];
+  }
+  let mode = 0;
+  for (let i = 0; i < 8; i++) {
+    if ((bitfield & 1) === 0) {
+      bitfield = bitfield >> 1;
+      break;
+    }
+    bitfield = bitfield >> 1;
+    mode++;
+  }
+  // If the mode is greater than 7, this is not a valid v1 skylink.
+  if (mode > 7) {
+    return [0n, 0n, 0n, "provided skylink has an invalid v1 bitfield"];
+  }
+
+  // Determine the offset and fetchSize increment.
+  const offsetIncrement = 4096 << mode;
+  let fetchSizeIncrement = 4096;
+  let fetchSizeStart = 0;
+  if (mode > 0) {
+    fetchSizeIncrement = fetchSizeIncrement << (mode - 1);
+    fetchSizeStart = (1 << 15) << (mode - 1);
+  }
+
+  // The next three bits decide the fetchSize.
+  let fetchSizeBits = bitfield & 7;
+  fetchSizeBits++; // semantic upstep, range should be [1,8] not [0,8).
+  const fetchSize = fetchSizeBits * fetchSizeIncrement + fetchSizeStart;
+  bitfield = bitfield >> 3;
+
+  // The remaining bits determine the offset.
+  const offset = bitfield * offsetIncrement;
+  if (offset + fetchSize > 1 << 22) {
+    return [0n, 0n, 0n, "provided skylink has an invalid v1 bitfield"];
+  }
+
+  // Return what we learned.
+  return [BigInt(version), BigInt(offset), BigInt(fetchSize), null];
+}
+
+// skylinkV1Bitfield sets the bitfield of a V1 skylink. It assumes the version
+// is 1 and the offset is 0. It will determine the appropriate fetchSize from
+// the provided dataSize.
+function skylinkV1Bitfield(dataSizeBI: bigint): [Uint8Array, string | null] {
+  // Check that the dataSize is not too large.
+  if (dataSizeBI > 1 << 22) {
+    return [new Uint8Array(0), "dataSize must be less than the sector size"];
+  }
+  const dataSize = Number(dataSizeBI);
```

Comment:
why do we expect bigint if this is going to be parsed as number right away tho ?

Reply:
all the number types in libskynet are bigints, it makes it compatible with all the other functions.